lib/types/jtd-schema.ts

```diff
@@ -0,0 +1,103 @@
+/** required keys of an object, not undefined */
+type RequiredKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? never : K
+}[keyof T]
+
+/** optional or undifined-able keys of an object */
+type OptionalKeys<T> = {
+  [K in keyof T]-?: undefined extends T[K] ? K : never
+}[keyof T]
+
+/** type is true if all elements of T extend T */
+type AllExtend_<T, E> = T extends E ? true : false
+type AllExtend<T, E> = false extends AllExtend_<T, E> ? false : true
+
+/** type is true if type is identically string */
+type IsString_<T> = T extends string ? (string extends T ? true : false) : false
+type IsString<T> = false extends IsString_<T> ? false : true
+
+/** gets only the string literals of a type or null if a type isn't a string literal */
+type NullableStringLiterals<T> = T extends string ? (string extends T ? null : T) : null
```

Comment:
maybe worth renaming to EnumString - shorter and connects to enum in schema - took me some time to "compute" it...

Reply:
That seems like a reasonable name, I called it `Nullable` because I'm kind of abusing the type and using the sentinel null to indicate anything that's not a string literal. This is necessary because we need to reject the type if there are any nulls present, i.e. it's not only a union of string literals. Still, I think EnumString convey this roughly and it's an internal name, so calling out the connection is good.

Reply:
NullableEnumString is also ok - up to you, just so it's clearer that it is related to enum.